object/src/mem.rs

```diff
@@ -728,8 +745,8 @@ where
                     let vr = e.vr();
                     // replace element
                     *e = DataElement::empty(tag, vr);
+                    self.len = Length::UNDEFINED;
```

Comment:
If I see this correctly, actions `SetStr`, `Set`, and the various `Push*` might also modify data without calling `put` (see e.g. line 811), so we'd also need some length resets in `apply_push_*_impl` and `apply_change_value_impl`.